### PR TITLE
Source schema to match source.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

This is the culmination of a load of small PRs.  It's switching the way KSQL deals with having `ROWTIME` and `ROWKEY` in the value schema.

Before this PR the sources, (streams/tables), stored in the Metastore has a value schema that included `ROWTIME` and `ROWKEY` fields, even though the value data in the source's topics do _not_ have such fields.  This contrasts to the internal topics for repartitioning and changelogs, which _do_ have `ROWTIME` and `ROWKEY` in the value data.  

The code around adding/removing these implicit fields was confusing and many many tests didn't set up sources with the correct schemas, leading to more confusion.

Some precursor PRs to this one fixed up the bad tests and simplified some of the logic, laying the way for this PR, which removes the implicit fields from the value schema of sources in the metastore.  So the schema of a `KsqlStream` or `KsqlTable` now matches the data in the source's topic.

Now that a source's schema matches the schema of the source's data the adding and removing of `ROWTIME` and `ROWKEY` to/from the value schemas has been limited to:

* `DataSourceNode`: It has always been this node that adds `mapValues` and `transformValues` steps to the KS topology to copy `ROWKEY` and `ROWTIME` to the value being passed around the KS topology. This has not changed, but now this node is also responsible for building the value schema to match.

* `KsqlStructuredDataOutputNode`: has always removed any field from the value named `ROWTIME` or `ROWKEY` by capturing the index of such fields in the value schema and passing them to `KStream.into`, which then filters them out.  This has not changed, but now the node is also responsible for removing the fields from the value schema.  

For better understanding of the current functionality, it's also worth calling out here the anomaly removing `ROWKEY` and `ROWTIME` in `KsqlStructuredDataOutputNode` causes:

* If you do `CREATE STREAM x AS SELECT * FROM y;` it will output values _without_ `ROWTIME `and `ROWKEY` and
* If you do `CREATE STREAM x AS SELECT * FROM y JOIN z ...` then it will output values _with_ both source's `ROWKEY` and `ROWTIME`.
* If you do `CREATE STREAM x AS SELECT ROWTIME, ROWKEY FROM x` then it will create an empty value schema. 

Note: This is existing functionality and does not change with this PR, though the PR adds tests to cover this. These same tests pass on the 5.3 branch.

### Testing done 

Suitable tests added to cover edge cases, but otherwise changes tested by current tests.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-functional-tests/src/test/resources/query-validation-tests/elements.json [hunk 1]
```diff
@@ -107,6 +107,55 @@
           "partitions": 4
         }
       ]
+    },
+    {
+      "name": "non-join with only ROWKEY and ROWTIME in projection",
+      "statements": [
+        "CREATE STREAM INPUT (F0 INT) WITH (kafka_topic='input', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT ROWTIME, ROWKEY FROM INPUT;"
```
Comment (agavra) [inline ksql-functional-tests/src/test/resources/query-validation-tests/elements.json]: can we add a test for `ROWTIME AS FOO` and make sure it shows up in the value?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/DecimalSumKudaf.java [hunk 2]
```diff
@@ -76,4 +76,8 @@ public BigDecimal undo(final BigDecimal valueToUndo, final BigDecimal aggregateV
     }
     return aggregateValue.subtract(valueToUndo, context);
   }
+
+  private static BigDecimal initialValue() {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/DecimalSumKudaf.java]: note: fixes check style error that is reporting in IDE only.

File: ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java [hunk 3]
```diff
@@ -281,18 +283,23 @@ private KeyField findKeyField(final List<SelectExpression> selectExpressions) {
...
           if (SchemaUtil.matchFieldName(keyField, qualifiedNameReference.getName().getSuffix())) {
-            return Optional.of(new Field(toName, i, keyField.schema()));
+            found = Optional.of(new Field(toName, i, keyField.schema()));
+            break;
           }
         }
       }
-      return Optional.empty();
+
+      return found
```
Comment (rodesai) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: when is it possible for either of these to be true?

Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java]: `GROUP BY ROWKEY` and `GROUP BY ROWTIME` if I remember rightly. Both of which work in 5.3.

**Comment by big-andy-coates** [comment]
Hi @hjafarpour,

This PR does not change the functionality you're talking about.  Before this PR any `ROWTIME` or `ROWKEY` field in a project is removed by the `KsqlStructuredOutputNode`.  This PR adds tests that cover this somewhat strange behaviour, but it does not change it. 

Whether these fields should be copied to the value schema or not is outside the scope of this PR and, IMHO, should be left as-is until we come to look at changing the query semantics, as was discussed with Matthias and everyone.

**Comment by big-andy-coates** [comment]
I'm merging this without a second +1 as I've waiting a week. I'm also merging even though @hjafarpour requested changes as he's his objections are unrelated to this PR and clearly explained in the description of this PR.

**Pull request merged**